File: CSharp/Proto/ToProtobuf.cs
#region

using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using ProtoBuf;

#endregion

namespace ThingModel.Proto
{
    public class ToProtobuf
    {
        // String declarations dictionary, for the sender side
        protected readonly IDictionary<string, int> StringDeclarations = new Dictionary<string, int>();

        // Current transaction object
        // It will be filled step by step
        protected Transaction Transaction;

        // Simple binding between ThingModel property types and protocol buffer property types
        // Tested as alternative to the dynamic keyword
        private readonly Dictionary<Type, PropertyType.Type> _prototypesBinding = new Dictionary
            <Type, PropertyType.Type>
            {
                {typeof (ThingModel.Property.Location), PropertyType.Type.LOCATION},
                {typeof (ThingModel.Property.String), PropertyType.Type.STRING},
                {typeof (ThingModel.Property.Double), PropertyType.Type.DOUBLE},
                {typeof (ThingModel.Property.Int), PropertyType.Type.INT},
                {typeof (ThingModel.Property.Boolean), PropertyType.Type.BOOLEAN},
                {typeof (ThingModel.Property.DateTime), PropertyType.Type.DATETIME},
            };

        // When a string is in this collection, it should be sent
        // as StringDeclaration for the next transaction
        private readonly HashSet<string> _stringToDeclare = new HashSet<string>();

        private readonly Dictionary<int, Thing> _thingsState = new Dictionary<int, Thing>();
        private readonly Dictionary<Tuple<int, int>, Property> _propertiesState = new Dictionary<Tuple<int, int>, Property>();

        protected int StringToKey(string value)
        {
            if (String.IsNullOrEmpty(value))
            {
                return 0;
            }

            int key;
            if (StringDeclarations.TryGetValue(value, out key))
            {
                return key;
            }

            key = StringDeclarations.Count + 1;

            StringDeclarations[value] = key;

            Transaction.string_declarations.Add(new StringDeclaration
                {
                    key = key,
                    value = value
                });

            return key;
        }

        public Transaction Convert(IEnumerable<ThingModel.Thing> publish, IEnumerable<ThingModel.Thing> delete,
                                   IEnumerable<ThingModel.ThingType> declarations, string senderID)
        {
            Transaction = new Transaction();

            Transaction.string_sender_id = StringToKey(senderID);

            ConvertDeleteList(delete);

            foreach (var thing in publish)
            {
                ConvertThing(thing);
            }

            ConvertDeclarationList(declarations);

            return Transaction;
        }


        private readonly MemoryStream _memoryInput = new MemoryStream();

        public byte[] Convert(Transaction transaction)
        {
            _memoryInput.SetLength(0);
            Serializer.Serialize(_memoryInput, transaction);
            return _memoryInput.ToArray();
        }

        protected void ConvertThing(ThingModel.Thing thing)
        {
            var change = false;

            var publication = new Thing
                {
                    string_id = StringToKey(thing.ID),
                    string_type_name = thing.Type != null ? StringToKey(thing.Type.Name) : 0,
                    connections_change = false
                };
            
            Thing previousThing;
            _thingsState.TryGetValue(publication.string_id, out previousThing);

            if (previousThing == null)
            {
                change = true;
            }

	        IList<ThingModel.Thing> connectedThingsCache = null;

            if ((previousThing == null && thing.ConnectedThingsCount > 0)
                || (previousThing != null && previousThing.connections.Count != thing.ConnectedThingsCount))
            {
                publication.connections_change = true;
            }
            else
            {
	            connectedThingsCache = thing.ConnectedThings;
                foreach (var connectedThing in connectedThingsCache)
                {
                    var connectionKey = StringToKey(connectedThing.ID);

                    if (previousThing == null || !previousThing.connections.Contains(connectionKey))
                    {
                        publication.connections_change = true;
                    }
                }
            }
            
            // If we don't have changes on the connection list,
            // it's useless to send it
            if (publication.connections_change)
            {
                change = true;
	            if (connectedThingsCache == null)
	            {
		            connectedThingsCache = thing.ConnectedThings;
	            }

                foreach (var connectedThing in connectedThingsCache)
                {
                    var connectionKey = StringToKey(connectedThing.ID);
                    publication.connections.Add(connectionKey);
                }
                
            }

            foreach (var property in thing.GetProperties())
            {
                var proto = new Property
                    {
                        string_key = StringToKey(property.Key)
                    };

                ConvertProperty((dynamic) property, proto);

                var key = new Tuple<int, int>(publication.string_id, proto.string_key);

                if (previousThing != null)
                {
                    Property previousProto;
                    if (_propertiesState.TryGetValue(key, out previousProto))
                    {
                        // ReSharper disable CompareOfFloatsByEqualityOperator
                        // Boring but necessary ?
                        if (previousProto.type == proto.type &&
                            previousProto.boolean_value == proto.boolean_value &&
                            previousProto.datetime_value == proto.datetime_value &&
                            previousProto.double_value == proto.double_value &&
                            previousProto.int_value == proto.int_value &&
                            ((previousProto.location_value == null && proto.location_value == null) ||
                             (previousProto.location_value != null && proto.location_value != null &&
                              previousProto.location_value.x == proto.location_value.x &&
                              previousProto.location_value.y == proto.location_value.y &&
                              previousProto.location_value.z == proto.location_value.z &&
                              previousProto.location_value.z_null == proto.location_value.z_null &&
                              previousProto.location_value.string_system == proto.location_value.string_system)) &&
                            ((previousProto.string_value == null && proto.string_value == null) ||
                             (previousProto.string_value != null && proto.string_value != null && ((
                              previousProto.string_value.string_value == proto.string_value.string_value &&
                              previousProto.string_value.value == proto.string_value.value) ||
                              (previousProto.string_value.value == property.ValueToString())))))
                        {
                            continue;
                        }

                        // ReSharper restore CompareOfFloatsByEqualityOperator
                                    
                    }

                    
                }

                change = true;
                
                publication.properties.Add(proto);
                _propertiesState[key] = proto;
            }

            if (change)
            {
                Transaction.things_publish_list.Add(publication);
                _thingsState[publication.string_id] = publication;
            }
        }

        protected void ConvertDeleteList(IEnumerable<ThingModel.Thing> publish)
        {
            foreach (var thing in publish)
            {
                var key = StringToKey(thing.ID);
                Transaction.things_remove_list.Add(key);
				ManageThingSuppression(key);
            }
        }

        protected void ConvertDeclarationList(IEnumerable<ThingModel.ThingType> declarations)
        {
            foreach (var thingType in declarations)
            {
                var declaration = new ThingType
                    {
                        string_name = StringToKey(thingType.Name),
                        string_description = StringToKey(thingType.Description)
                    };

                foreach (var propertyType in thingType.GetProperties())
                {
                    declaration.properties.Add(new PropertyType
                        {
                            string_key = StringToKey(propertyType.Key),
                            string_name = StringToKey(propertyType.Name),
                            string_description = StringToKey(propertyType.Description),
                            required = propertyType.Required,
                            type = _prototypesBinding[propertyType.Type]
                        });
                }

                Transaction.thingtypes_declaration_list.Add(declaration);
            }
        }

        protected void ConvertProperty(ThingModel.Property.Location property, Property proto)
        {
            var value = property.Value;

            if (value is Location.LatLng)
            {
                proto.type = Property.Type.LOCATION_LATLNG;
            }
            else if (value is Location.Point)
            {
                proto.type = Property.Type.LOCATION_POINT;
            }
            else
            {
                proto.type = Property.Type.LOCATION_EQUATORIAL;    
            }
            
            proto.location_value = new Property.Location
                {
                    x = value.X,
                    y = value.Y,
                    string_system = StringToKey(value.System),
                    z_null = value.Z == null
                };

            if (value.Z != null)
            {
                proto.location_value.z = (double) value.Z;
            }
        }

        protected void ConvertProperty(ThingModel.Property.String property, Property proto)
        {
            var value = property.Value;

            proto.type = Property.Type.STRING;
            proto.string_value = new Property.String();
            
            if (_stringToDeclare.Contains(value))
            {
                proto.string_value.string_value = StringToKey(value);
            }
            else
            {
                proto.string_value.value = value;

                // Use string to key the next time
                _stringToDeclare.Add(value);
            }
        }

        protected void ConvertProperty(ThingModel.Property.Double property, Property proto)
        {
            proto.type = Property.Type.DOUBLE;
            proto.double_value = property.Value;
        }

        protected void ConvertProperty(ThingModel.Property.Int property, Property proto)
        {
            proto.type = Property.Type.INT;
            proto.int_value = property.Value;
        }

        protected void ConvertProperty(ThingModel.Property.Boolean property, Property proto)
        {
            proto.type = Property.Type.BOOLEAN;
            proto.boolean_value = property.Value;
        }

        protected void ConvertProperty(ThingModel.Property.DateTime property, Property proto)
        {
            proto.type = Property.Type.DATETIME;
            proto.datetime_value = property.Value.Ticks;
        }

	    protected void ManageThingSuppression(int thingId)
	    {
		    _thingsState.Remove(thingId);

		    var toRemove = new List<Tuple<int, int>>();
		    foreach (var tuple in _propertiesState.Keys)
		    {
			    if (tuple.Item1 == thingId)
			    {
				    toRemove.Add(tuple);
			    }
		    }

		    foreach (var tuple in toRemove)
		    {
			    _propertiesState.Remove(tuple);
		    }
	    }

	    public void ApplyThingSuppressions(IEnumerable<ThingModel.Thing> things)
	    {
		    foreach (var thing in things)
		    {
				int key;
				if (StringDeclarations.TryGetValue(thing.ID, out key))
				{
					ManageThingSuppression(key);
				}
		    }
	    }
    }
}

File: TypeScript/Proto/ToProtobuf.ts
module ThingModel.Proto {

	export class ToProtobuf {
		// Current transaction object
		// it will be filled step by step
		private _transaction: Transaction;

		// String declaration dictionary
		public StringDeclarations: { [value: string]: number };
		public StringDeclarationsCpt: number;

		// String to declares
		private _stringToDeclare: { [value: string]: boolean };

		private _thingsState: { [id: number]: Thing };
		private _propertiesState: { [key: string]: Property };

		constructor() {
			this.StringDeclarations = {};
			this.StringDeclarationsCpt = 0;
			this._stringToDeclare = {};
			this._thingsState = {};
			this._propertiesState = {};
		}

		private StringToKey(value: string): number {
			if (!value) {
				return 0;
			}

			var key = this.StringDeclarations[value];

			if (key) {
				return key;
			}

			key = ++this.StringDeclarationsCpt;
			this.StringDeclarations[value] = key;

			var stringDeclaration = new ProtoTools.Builder.StringDeclaration();

			stringDeclaration.setKey(key);
			stringDeclaration.setValue(value);

			this._transaction.string_declarations.push(stringDeclaration);

			return key;
		}

		public Convert(publish: ThingModel.Thing[],
			deletions: ThingModel.Thing[],
			declarations: ThingModel.ThingType[],
			senderID: string): Transaction {

			this._transaction = new ProtoTools.Builder.Transaction();

			this._transaction.setStringSenderId(this.StringToKey(senderID));

			_.each(publish, (thing) => {
				this.ConvertThing(thing);
			});

			this.ConvertDeleteList(deletions);
			this.ConvertDeclarationList(declarations);

			return this._transaction;

		}

		public ConvertTransaction(transaction: Transaction): ArrayBuffer {
			return transaction.toArrayBuffer();
		}

		private ConvertThing(thing: ThingModel.Thing): void {
			var change = false;

			var thingID = this.StringToKey(thing.ID);

			var publication = new ProtoTools.Builder.Thing();
			publication.setStringId(thingID);
			publication.setStringTypeName(thing.Type != null ?
				this.StringToKey(thing.Type.Name) : 0);
			publication.setConnectionsChange(false);

			var previousThing = this._thingsState[thingID];

			if (previousThing == null || previousThing.getStringTypeName()
				!= publication.getStringTypeName()) {
				change = true;
			}

			var connectedThingsCache: ThingModel.Thing[] = null;

			if ((previousThing == null && thing.ConnectedThingsCount > 0)
				|| (previousThing != null && previousThing.getConnections().length !=
					thing.ConnectedThingsCount)) {
				publication.setConnectionsChange(true);
			} else {
				connectedThingsCache = thing.ConnectedThings;

				_.any(connectedThingsCache, (connectedThing: ThingModel.Thing) => {
					var connectionKey = this.StringToKey(connectedThing.ID);

					if (previousThing == null || !_.contains(previousThing.getConnections(), connectionKey)) {
						publication.setConnectionsChange(true);
						return true;
					}
					return false;
				});
			}

			if (publication.getConnectionsChange()) {
				change = true;

				if (!connectedThingsCache) {
					connectedThingsCache = thing.ConnectedThings;
				}

				_.each(connectedThingsCache, (connectedThing: ThingModel.Thing) => {
					var connectionKey = this.StringToKey(connectedThing.ID);
					publication.connections.push(connectionKey);
				});
			}

			_.each(thing.Properties, (property: ThingModel.Property) => {
				var propertyId = this.StringToKey(property.Key);

				var proto = new ProtoTools.Builder.Property();
				proto.setStringKey(propertyId);

				switch (property.Type) {
				case Type.LocationLatLng:
				case Type.LocationPoint:
				case Type.LocationEquatorial:
					this.ConvertLocationProperty(property, proto);
					break;
				case Type.String:
					this.ConvertStringProperty(
						<ThingModel.Property.String> property, proto);
					break;
				case Type.Double:
					proto.setType(Property.Type.DOUBLE);
					proto.setDoubleValue(
					(<ThingModel.Property.Double> property).Value);
					break;
				case Type.Int:
					proto.setType(Property.Type.INT);
					proto.setIntValue(
					(<ThingModel.Property.Int> property).Value);
					break;
				case Type.Boolean:
					proto.setType(Property.Type.BOOLEAN);
					proto.setBooleanValue(
					(<ThingModel.Property.Boolean> property).Value);
					break;
				case Type.DateTime:
					proto.setType(Property.Type.DATETIME);
					proto.setDatetimeValue(
					(<ThingModel.Property.DateTime> property).Value.getTime());
					break;
				}

				// We are in a JavaScript world :-)
				var propertyStateKey = thingID + ":" + propertyId;

				if (previousThing != null) {
					var previousProto = this._propertiesState[propertyStateKey];

					// Terrible part :)
					if (previousProto != null &&
						previousProto.getType() == proto.getType() &&
						previousProto.getBooleanValue() == proto.getBooleanValue() &&
						previousProto.getDatetimeValue() == proto.getDatetimeValue() &&
						previousProto.getDoubleValue() == proto.getDoubleValue() &&
						previousProto.getIntValue() == proto.getIntValue()) {
						var previousLoc = previousProto.getLocationValue();
						var loc = proto.getLocationValue();

						if ((previousLoc == null && loc == null) ||
						(previousLoc != null && loc != null &&
							previousLoc.x == loc.x &&
							previousLoc.y == loc.y &&
							previousLoc.z == loc.z &&
							previousLoc.z_null == loc.z_null &&
							previousLoc.getStringSystem() == loc.getStringSystem())) {
							var previousStr = previousProto.getStringValue();
							var str = proto.getStringValue();

							if ((previousStr == null && str == null) ||
							(previousStr != null && str != null &&
							((previousStr.getStringValue() == str.getStringValue() &&
								previousStr.getValue() == str.getValue()) ||
							(property.ValueToString() == previousStr.getValue())))) {
								return;
							}
						}
					}
				}

				change = true;
				publication.properties.push(proto);
				this._propertiesState[propertyStateKey] = proto;
			});

			if (change) {
				this._transaction.things_publish_list.push(publication);
				this._thingsState[thingID] = publication;
			}
		}

		private ConvertDeleteList(list: ThingModel.Thing[]): void {
			_.each(list, (thing: ThingModel.Thing) => {
				var key = this.StringToKey(thing.ID);
				this._transaction.things_remove_list.push(key);

				this.ManageThingSuppression(key);
			});
		}

		private ConvertDeclarationList(list: ThingModel.ThingType[]): void {
			_.each(list, (thingType: ThingModel.ThingType) => {
				var declaration = new ProtoTools.Builder.ThingType();

				declaration.setStringName(this.StringToKey(thingType.Name));
				declaration.setStringDescription(this.StringToKey(thingType.Description));

				_.each(thingType.Properties, (propertyType: ThingModel.PropertyType) => {
					var prop = new ProtoTools.Builder.PropertyType();
					prop.setStringKey(this.StringToKey(propertyType.Key));
					prop.setStringName(this.StringToKey(propertyType.Name));
					prop.setStringDescription(this.StringToKey(propertyType.Description));
					prop.setRequired(propertyType.Required);

					switch (propertyType.Type) {
					case Type.LocationLatLng:
					case Type.LocationEquatorial:
					case Type.LocationPoint:
					case Type.Location:
						prop.setType(PropertyType.Type.LOCATION);
						break;
					case Type.String:
						prop.setType(PropertyType.Type.STRING);
						break;
					case Type.Double:
						prop.setType(PropertyType.Type.DOUBLE);
						break;
					case Type.Int:
						prop.setType(PropertyType.Type.INT);
						break;
					case Type.Boolean:
						prop.setType(PropertyType.Type.BOOLEAN);
						break;
					case Type.DateTime:
						prop.setType(PropertyType.Type.DATETIME);
						break;
					}

					declaration.properties.push(prop);
				});

				this._transaction.thingtypes_declaration_list.push(declaration);
			});
		}

		private ConvertLocationProperty(property: ThingModel.Property,
			proto: Property) {

			var value = (<any>property).Value;

			switch (value.type) {
			case "latlng":
				proto.setType(Property.Type.LOCATION_LATLNG);
				break;
			case "equatorial":
				proto.setType(Property.Type.LOCATION_EQUATORIAL);
				break;
			case "point":
			default:
				proto.setType(Property.Type.LOCATION_POINT);
				break;

			}

			var loc = new ProtoTools.Builder.Property.Location();

			loc.setX(value.X);
			loc.setY(value.Y);
			loc.setStringSystem(this.StringToKey(value.System));
			loc.setZNull(value.Z == null);

			if (value.Z != null) {
				loc.setZ(value.Z);
			}

			proto.setLocationValue(loc);
		}

		private ConvertStringProperty(property: ThingModel.Property.String,
			proto: Property) {
			var value = property.Value;
			proto.setType(Property.Type.STRING);

			var st = new ProtoTools.Builder.Property.String();
			if (value && this._stringToDeclare[value]) {
				st.setStringValue(this.StringToKey(value));
			} else {
				st.setValue(value);

				// Use a string decleration next time
				this._stringToDeclare[value] = true;
			}

			proto.setStringValue(st);
		}

		private ManageThingSuppression(thingId: number): void {
			delete this._thingsState[thingId];

			var stringId = thingId + ":";

			_.each(this._propertiesState, (value, key) => {
				if (key.indexOf(stringId) === 0) {
					delete this._propertiesState[key];
				}
			});
		}

		public ApplyThingsSuppressions(things: ThingModel.Thing[]): void {
			_.each(things, (thing: ThingModel.Thing) => {
				var key = this.StringDeclarations[thing.ID];

				if (key) {
					this.ManageThingSuppression(key);
				}
			});
		}
	}

}
